spec/forms/user_piv_cac_login_form_spec.rb

```diff
@@ -0,0 +1,104 @@
+  end
+  let(:nonce) { 'nonce' }
+  let(:piv_cac_required) { false }
+  let(:form) { described_class.new(token: token, nonce: nonce, piv_cac_required: piv_cac_required) }
+
+  describe '#submit' do
+    before(:each) do
+      allow(PivCacService).to receive(:decode_token).with(token) { token_response }
+    end
+
+    context 'when token is valid' do
+      let(:token) { 'good-token' }
+
+      let(:token_response) do
+        {
+          'uuid' => piv_cac_configuration.x509_dn_uuid,
+          'subject' => 'x509-subject',
+          'nonce' => nonce,
+        }
+      end
+
+      it 'returns FormResponse with success: true' do
+        result = form.submit
+
+        expect(result.success?).to eq true
+        expect(result.errors).to eq({})
+        expect(result.extra).to eq({ key_id: nil })
+      end
+    end
+
+    context 'when token is invalid' do
+      let(:token) { 'bad-token' }
+      let(:token_response) do
+        { 'error' => 'token.bad', 'nonce' => nonce, 'key_id' => 'foo' }
+      end
+
+      it 'returns FormResponse with success: false' do
+        result = form.submit
+
+        expect(result.success?).to eq false
+        expect(result.errors).to eq({ type: 'token.bad' })
+        expect(result.extra).to eq({ key_id: 'foo' })
+      end
+    end
+
+    context 'when nonce is invalid' do
+      let(:token) { 'bad-token' }
+      let(:token_response) do
+        { 'error' => 'token.bad', 'nonce' => bad_nonce, 'key_id' => 'foo' }
+      end
+      let(:bad_nonce) { nonce + 'X' }
+
+      it 'returns FormResponse with success: false' do
+        result = form.submit
+
+        expect(result.success?).to eq false
+        expect(result.errors).to eq({ type: 'token.invalid' })
+        expect(result.extra).to eq({ key_id: 'foo' })
+      end
+    end
+
+    context 'when piv cac is required' do
+      let(:token) { 'good-token' }
+      let(:piv_cac_required) { true }
+
+      it 'returns FormResponse with success: true when the token indicates auth cert' do
+        resp = { 'nonce' => nonce, 'is_auth_cert' => true, 
+                 'uuid' => piv_cac_configuration.x509_dn_uuid }
+        allow(PivCacService).to receive(:decode_token).with(token) { resp }
+
+        result = form.submit
+
+        expect(result.success?).to eq true
+      end
+
+      it 'returns FormResponse with success: false when the token indicates not an auth cert' do
+        resp = { 'nonce' => nonce, 'is_auth_cert' => false, 
+                 'uuid' => piv_cac_configuration.x509_dn_uuid }
+        allow(PivCacService).to receive(:decode_token).with(token) { resp }
+
+        result = form.submit
+
+        expect(result.success?).to eq false
+        expect(result.errors).to eq({ type: 'certificate.not_auth_cert' })
+      end
+    end
+
+    context 'when token is missing' do
+      let(:token) {}
+
+      it 'returns FormResponse with success: false' do
+        result = form.submit
+
+        expect(result.success?).to eq false
+      end
```

Comment:
this is the failing spec for the old behavior right?

Reply:
This is a kind of weird case where a failing spec was difficult to write since the previous form behavior didn't have a way to know whether HSPD12 was required.

https://github.com/18F/identity-idp/pull/5125/files#diff-3bb8729bc0e25198ab1d5fd037f2f5b2a00f0accd5ff4dfacf892b0bb16cbc1eR82-R91 is the closest to the failing spec though.